### ISSUE TITLE
error if coordinates are invalid

### ISSUE THREAD

**Issue by contra** [tidwall/tile38#361]
Continuing discussion from https://github.com/tidwall/tile38/issues/328

```json
{"type":"Point","coordinates":[30.60775148298956,-96.3439205355819]}
```

If I were to put a boundary for earth in tile38 (full extent of the WGS84 range) tile38 would return 0 results when I asked for an intersection.

I think the correct behavior is to error, because this isn't a valid coordinate and tile38 uses WGS84. I think this can be applied comprehensively anywhere coordinate inputs are handled. This enables us to relay this to the end-user as a 400 (you sent us a bad coordinate, polygon, whatever).

**Comment by tidwall** [comment]
In the section about Bounding Boxes, the [GeoJSON spec](https://tools.ietf.org/html/rfc7946#section-3.1.9) does state: 

> Implementers MUST NOT use latitude values greater than 90 or less than -90 to imply an extent that is not a spherical cap.

So an error is probably the best long term solution. 

In the meantime it should be possible to query beyond the -/+180,90 extent.

**Comment by stevelacy** [comment]
@tidwall Would you be open to accepting a PR that adds coordinate validation that validates based on WGS84 with an opt-in option?

**Comment by tidwall** [comment]
I'm cool with that. Were you thinking to just validate the lat/lons for -/+180,90, and throw an error?

**Comment by stevelacy** [comment]
Yes, I'd check if the lat/lons were not 0 and -/+ 180,90.

**Comment by contra** [comment]
@tidwall Coordinate validation is an easy first step but could go further, checking for self-intersections/geojson spec compliance/other issues.

Prior art:
- https://postgis.net/docs/ST_IsValid.html
- https://postgis.net/docs/ST_IsValidReason.html
  - (Notes about some checks it does: https://postgis.net/docs/using_postgis_dbmanagement.html#OGC_Validity).
- gdal geometry validation
- https://github.com/mapbox/geojsonhint

**Comment by tidwall** [comment]
I agree. Perhaps the easiest approach is to add a `Valid` method to the GeoJSON `Object` type.

https://github.com/tidwall/geojson/blob/master/object.go#L31-L45

Also having a parse option list 'RequireValid=true'

https://github.com/tidwall/geojson/blob/master/object.go#L75-L90

**Comment by tidwall** [comment]
Then from Tile38, maybe have a configuration option like `require-valid` ?

**Comment by stevelacy** [comment]
Sounds good, perhaps in the future we could add a validate command to validate that the input geojson is valid? `tile38-cli > validate "{geojson}"`

**Comment by tidwall** [comment]
I’d consider adding a VALIDATE in the future. 
But am I correct that right now you are wanting the validate to automatically happen on SET, as long as the config option is active?

**Comment by tidwall** [comment]
Makes sense. I say go for it. I think it'll be a nice enhancement.
